kong/init.lua

```diff
@@ -408,40 +606,92 @@ function Kong.access()
 
   ctx.delay_response = false
 
-  core.access.after(ctx)
+  core.access.after(ctx, mock)
 end
 
-function Kong.header_filter()
+function Kong.header_filter(mock)
   local ctx = ngx.ctx
-  core.header_filter.before(ctx)
+  core.header_filter.before(ctx, mock)
 
-  for plugin, plugin_conf in plugins_iterator(singletons.loaded_plugins) do
-    plugin.handler:header_filter(plugin_conf)
+  if mock then
+    if ctx.KONG_PROXIED then
+      mock_plugins(ctx)
+
+      ngx.header.content_length = nil
+      ngx.header.content_type   = "application/json"
+    end
+
+  else
+    for plugin, plugin_conf in plugins_iterator(singletons.loaded_plugins) do
+      plugin.handler:header_filter(plugin_conf)
+    end
   end
 
-  core.header_filter.after(ctx)
+  core.header_filter.after(ctx, mock)
 end
 
-function Kong.body_filter()
+function Kong.body_filter(mock)
   local ctx = ngx.ctx
-  for plugin, plugin_conf in plugins_iterator(singletons.loaded_plugins) do
-    plugin.handler:body_filter(plugin_conf)
+
+  if mock then
+    if ctx.KONG_PROXIED then
+      mock_plugins(ctx)
+
+      local ctx = ngx.ctx
+      local mock = ctx.mock
+
+      mock.request = (mock.request or "") .. ngx.arg[1]
+
+      if ngx.arg[2] then
+        mock_plugins(ctx, "log")
+
+        local cjson = require "cjson.safe"
+        mock.request = cjson.decode(mock.request)
+        ngx.arg[1] = cjson.encode(mock)
+
+      else
+        ngx.arg[1] = nil
+      end
+    end
+
+  else
+    for plugin, plugin_conf in plugins_iterator(singletons.loaded_plugins) do
+      plugin.handler:body_filter(plugin_conf)
+    end
   end
 
-  core.body_filter.after(ctx)
+  core.body_filter.after(ctx, mock)
 end
 
-function Kong.log()
+function Kong.log(mock)
   local ctx = ngx.ctx
-  for plugin, plugin_conf in plugins_iterator(singletons.loaded_plugins) do
-    plugin.handler:log(plugin_conf)
+
+  if not mock then
+    for plugin, plugin_conf in plugins_iterator(singletons.loaded_plugins) do
+      plugin.handler:log(plugin_conf)
+    end
   end
 
-  core.log.after(ctx)
+  core.log.after(ctx, mock)
 end
 
-function Kong.handle_error()
-  return kong_error_handlers(ngx)
+function Kong.mock_service()
```

Comment:
style: need 2 line jumps (saw a couple other similar places that needed this fix)

Reply:
I'm not sure if it makes sense to proxy to a stub endpoint when using mock... We are trying to get visibility into Kong's behavior (as in, into the Lua VM, plugins, and choices it makes). If one wants to debug the _contents_ of the proxied data (e.g. what the request looks like coming into Kong vs. out), then I believe it is a different kind of debugging, that does not necessitate the `mock` server?